nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,294 @@
+#include <boost/optional.hpp>
+#include <boost/polymorphic_pointer_cast.hpp>
+#include <nano/lib/numbers.hpp>
+#include <nano/lib/timer.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/confirmation_height_processor.hpp>
+#include <nano/node/node.hpp> // For active_transactions
+#include <nano/secure/blockstore.hpp>
+#include <nano/secure/common.hpp>
+#include <nano/secure/ledger.hpp>
+
+constexpr std::chrono::milliseconds nano::confirmation_height_processor::batch_write_delta;
+
+nano::confirmation_height_processor::confirmation_height_processor (nano::pending_confirmation_height & pending_confirmation_height, nano::block_store & store, nano::ledger & ledger, nano::active_transactions & active, nano::logger_mt & logger) :
+pending_confirmations (pending_confirmation_height),
+store (store),
+ledger (ledger),
+active (active),
+logger (logger),
+thread ([this]() {
+	nano::thread_role::set (nano::thread_role::name::confirmation_height_processing);
+	this->run ();
+})
+{
+}
+
+nano::confirmation_height_processor::~confirmation_height_processor ()
+{
+	stop ();
+}
+
+void nano::confirmation_height_processor::stop ()
+{
+	stopped = true;
+	condition.notify_one ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+void nano::confirmation_height_processor::run ()
+{
+	std::unique_lock<std::mutex> lk (pending_confirmations.mutex);
+	while (!stopped)
+	{
+		if (!pending_confirmations.queue.empty ())
+		{
+			auto pending_confirmation = pending_confirmations.queue.front ();
+			pending_confirmations.queue.pop ();
+			lk.unlock ();
+			add_confirmation_height (pending_confirmation);
+			lk.lock ();
+		}
+		else
+		{
+			condition.wait (lk);
+		}
+	}
+}
+
+void nano::confirmation_height_processor::add (nano::block_hash const & hash_a)
+{
```

Comment:
We could store if a block is live here which is before the potentially long delay of updating confirmation height.

Reply:
This is only called for blocks which are from the live network/have `block_confirm` called on them.